client/state/connectionMiddleware/middleware.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * External dependencies
+ */
+import uniqueId from 'lodash/utility/uniqueId';
+
+/**
+ * Internal dependencies
+ */
+import {
+	isOffline,
+	isOnline
+} from 'state/application/selectors';
+import {
+	OFFLINE_QUEUE_ACTION,
+	CONNECTION_RESTORED,
+	OFFLINE_QUEUE_REMOVE,
+} from 'state/action-types';
+const ACTION_TIMEOUT = 10000;
+
+const connections = {};
+
+let actionTimeout = null;
+
+//Prob should be in actions
+function queueAction( action ) {
+	return {
+		type: OFFLINE_QUEUE_ACTION,
+		id: uniqueId(),
+		dispatchedAt: ( new Date() ).getTime(),
+		action: action
+	}
+}
+
+//Prob should be in actions
+function removeAction( id ) {
+	return {
+		type: OFFLINE_QUEUE_REMOVE,
+		id: id
+	}
+}
+
+//Prob should be in selectors
+function getFirstAction( state ) {
+	return state.connectionMiddleware.actionQueue[ 0 ];
+}
+
+/**
+ *  Fires next action from the queue.
+ *  Should be called by a wrapped dispatchAndFireNextQueuedAction.
+ *  But in case the connectionFunction fails or does not dispatch anything,
+ *  debouncer kicks in.
+ */
+function fireNextQueuedAction( store ) {
+	clearTimeout( actionTimeout );
+	const firstAction = getFirstAction( store.getState() );
+
+	if ( !isOffline( store.getState() ) && firstAction ) {
+		//for testing:
+		firstAction.action.noQueue = true;
+
+		store.dispatch( removeAction( firstAction.id ) );
+		setTimeout( fireNextQueuedAction.bind( null, store ), ACTION_TIMEOUT );
```

Comment:
What happens when a dispatched action succeeds after the timeout? We may need some extra state here so it knows it's not in process any more and shouldn't trigger the next dispatch.